### PR TITLE
feat(discard): add new cell_emitter feature with cell filter feature on and he…

### PR DESCRIPTION
**Pull request by ashuralyk**
## background
Forcerelay should integrate `cell-emitter` to enable CKB header and cell sync feature

## todo list
- [x] develop cell filter feature
- [x] develop header sync feature
- [x] optimize Axon's monitor to test the disconnection of WebSocket
- [x] optimize CKB's monitor to solve the missing of channel_id while building packet tx
- [x] test with `cell-emitter` and Axon

## related PRs
- https://github.com/synapseweb3/ibc-solidity-contract/pull/33

## need further discussion
Axon seems to change its security assumption on integration of `cell-emitter` about the header and cell relaying @Flouse

### CODE REVIEW

**Comment by ashuralyk** [comment]
@Flouse  this PR can start to review

File: crates/relayer/tests/config/fixtures/relayer_conf_example.toml [hunk 1]
```diff
@@ -94,8 +94,9 @@ transfer_contract_address = "0x0000000000000000000000000000000000000000"
 restore_block_count = 10000
 key_name = "relayer_axon_wallet"
 store_prefix = "forcerelay"
-ckb_light_client_contract_address = "0x71C7656EC7ab88b098defB751B7401B5f6d8976F"
-image_cell_contract_address = "0x71C7656EC7ab88b098defB751B7401B5f6d8976F"
+emitter_ckb_url = "https://testnet.ckbapp.dev"
+emitter_cell_start_block_number = 1000000
+emitter_header_start_block_number = 1000000
```
Comment (Flouse) [inline crates/relayer/tests/config/fixtures/relayer_conf_example.toml:99]: why did you choose `1000000` as the default value and not 10,000,000?

File: crates/relayer/src/chain/axon/emitter.rs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+use super::ContractProvider;
+
+const fn system_contract_address(addr: u8) -> Address {
+    H160([
+        0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+        0xff, 0xff, 0xff, 0xff, addr,
+    ])
+}
+
```
Comment (Flouse) [inline crates/relayer/src/chain/axon/emitter.rs:32]:
```suggestion
/// see https://docs.axonweb3.io/contract/system_contacts
const fn system_contract_address(addr: u8) -> Address {
    H160([
        0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
        0xff, 0xff, 0xff, 0xff, addr,
    ])
}

```

File: crates/relayer/src/chain/axon/monitor.rs [hunk 3]
```diff
@@ -112,22 +118,21 @@ impl AxonEventMonitor {
...
             )))?;
-        let event_filter = |event: &OwnableIBCHandlerEvents| {
+        let event_filter = |event: &ContractEvents| {
             matches!(
                 event,
-                OwnableIBCHandlerEvents::SendPacketFilter(_)
-                    | OwnableIBCHandlerEvents::WriteAcknowledgementFilter(_)
+                ContractEvents::SendPacketFilter(_)
+                    | ContractEvents::WriteAcknowledgementFilter(_)
+                    | ContractEvents::RegisterCellEmitterFilterFilter(_)
```
Comment (blckngm) [inline crates/relayer/src/chain/axon/monitor.rs:134]: * How and when are these events emitted?
* What about previously registered filters?
* What's the significance of storing these filters on chain? (What's the difference with e.g. storing them offchain in a db?)
* Are all cells that we are interested in guaranteed to be emitted? Are they guaranteed to be emitted before the relavant ibc event is relayed?

File: crates/relayer/tests/config/fixtures/relayer_conf_example.toml (hunk 1 above)
Comment (ashuralyk) [reply crates/relayer/tests/config/fixtures/relayer_conf_example.toml:99]: it's just an example file for test and won't work for real world condition

File: crates/relayer/src/chain/axon/monitor.rs (hunk 3 above)
Comment (ashuralyk) [reply crates/relayer/src/chain/axon/monitor.rs:134]: here's the answers:
1. this event filter is set to filter events before starting Axon's monitor
2. previous filters can be found on-chain though search prior blocks of Axon, but only accept 25000 blocks ago, filters registered before this range cannot be replayed
3. filters are not stored on chain for current solidity implementation, and I don't suggest to bring off-chain db to relayer, it shouldn't keep status
4. yes, almost all of cells except AckPacket cell

**Review by jjyr — APPROVED** [review]

**Comment by ashuralyk** [comment]
@jjyr do you know why CI test between two Axon nodes taking so long that killed by Action

**Pull request closed without merging**